rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,274 @@
+making an administrative request.
+
+Pros:
+
+- Client can apply common logic for MFA verification, rather than endpoint
+  specific logic.
+  - Ex: First request without MFA, receive access denied error, re-request with MFA.
+  - MFA verification can be extended to additional Auth endpoint without any client changes.
+- Easily extends to other endpoints that always or sometimes need MFA verification.
+- Auth is completely handled by certificates rather than an augmented client request.
+
+Cons:
+
+- MFA verification applies to any requests using the MFA verified certificates.
+  This can be used as a MFA-replay-like attack since only one MFA verification is
+  required to carry out multiple administrative actions.
+  - Like Per-session MFA, this puts the security of the system more in the hands
+    of the client to not misuse or expose the certificates within their one
+    minute lifetime.
+- Requires users to create a new client with the reissued certificates, resulting
+  in additional connections to the Auth server.
+
+##### Option choice: TBD
+
+I am slightly in favor of option 2 due to the flexibility of the system compared
+to option 1 and the security benefit over option 3. However, we may favor the
+simplicity of defining MFA requirements on the gRPC layer (option 1) or of
+enforcing MFA based on the user's certificate alone (option 3).
+
+@Reviewers please offer your opinions and I will update this section once we
+decide on an option.
+
+#### Server changes
+
+For admin actions, the Auth server will validate MFA for each request, either
+from an `MFAChallengeResponse` passed by the client or an MFA-verified certificate
+used for the request.
+
+If the request fails MFA verification, an access denied
+error will be returned to the client.
+
+```go
+// ErrAPIMFARequired is returned by AccessChecker when an API request
+// requires an MFA check.
+var ErrAPIMFARequired = trace.AccessDenied("API request requires MFA")
+```
+
+The client will check for this error to determine whether it should retry a
+request with MFA verification.
+
+### Other considerations
+
+### Automated use cases
+
+If a user has no MFA device registered, they will not need to pass MFA
+verification for admin actions. This means that you can impersonate a user
+without an MFA device registered to get certificates for that user which bypass
+this MFA requirement. This is essential for automated use cases such as the
+[Teleport Terraform Provider](https://goteleport.com/docs/management/guides/terraform-provider/)
+or [Access Request Plugins](https://goteleport.com/docs/access-controls/access-request-plugins/).
+to continue to work as expected.
+
+#### Hardware Key Support
+
+Hardware Key Support can also be used for this use case. When setting
+`require_session_mfa: hardware_key_touch` for a role or cluster, YubiKey tap is
+required for _every_ private key operation, which applies to essentially every
+action you can take in Teleport.
```

Comment:
It's handled separately from MFA so it can not be used as a stand in replacement for MFA currently, if that's what you mean.  

If you are using `require_session_mfa: hardware_key`, then you must pass MFA in addition to get the `MFAVerified` extension and use a PIV key to get the `PrivateKeyPolicy=hardware_key` extension.

Reply:
Got it, thanks.